sqlitecluster/SQLite.cpp

```diff
@@ -208,18 +212,82 @@ void SQLite::_sqliteLogCallback(void* pArg, int iErrCode, const char* zMsg) {
+    } else {
+        // If we get here, then full checkpoints are enabled, and we have enough pages in the WAL file to perform one.
+        SINFO("[checkpoint] " << pageCount << " pages behind, beginning complete checkpoint.");
+
+        // This thread will run independently. We capture the variables we need here and pass them by value.
+        string filename = object->_filename;
+        string dbNameCopy = dbName;
+        thread([object, filename, dbNameCopy]() {
+            SInitialize("checkpoint");
+            uint64_t start = STimeNow();
+
+            // Lock the mutex that keeps anyone from starting a new transaction.
+            object->_sharedData->blockNewTransactionsMutex.lock();
+
+            while (1) {
+                // Lock first, this prevents anyone from updating the count while we're operating here.
+                unique_lock<mutex> lock(object->_sharedData->notifyWaitMutex);
+
+                // Now that we have the lock, check the count. If there are no outstanding transactions, we can
+                // checkpoint immediately, and then we'll return.
+                int count = object->_sharedData->currentTransactionCount.load();
+                SINFO("[checkpoint] Waiting on " << count << " remaining transactions.");
+
+                if (count == 0) {
+                    // Grab the global commit lock. Then we can look up this object and see if it still exists.
+                    // This is safe to do, we know nobody's committing, since we just waited for all transactions
+                    // to be finished. Why this global lock? Because we re-used it for modifying SharedData
+                    // objects, because that's only done at creation/destruction of SQLite objects and here.
+                    SQLITE_COMMIT_AUTOLOCK;
+
+                    // Verify the SQLite object passed into this function still exists. It's feasible (though
+                    // unlikely), that it could have been deleted if we tried to run a checkpoint just before
+                    // shutting down (or otherwise destroying an SQLite object).
+                    auto it = _sharedDataLookupMap.find(filename);
+                    if (it == _sharedDataLookupMap.end() || it->second->validObjects.find(object) == it->second->validObjects.end()) {
+                        SWARN("Aborting checkpoint, SQLite object deleted.");
+                        break;
+                    }
+
+                    // Time and run the checkpoint operation.
+                    uint64_t checkpointStart = STimeNow();
+                    SINFO("[checkpoint] Waited " << ((checkpointStart - start) / 1000)
+                          << "ms for pending transactions. Starting complete checkpoint.");
+                    int walSizeFrames = 0;
+                    int framesCheckpointed = 0;
+                    sqlite3_wal_checkpoint_v2(object->_db, dbNameCopy.c_str(), SQLITE_CHECKPOINT_RESTART, &walSizeFrames, &framesCheckpointed);
```

Comment:
Should we check the return status here? If we fail to do the full checkpoint it'll return SQLITE_BUSY instead of SQLITE_OK. I think we would want to know if this is the case, as it means we did a passive check point here instead of a FULL.